src/aapns/pool.py

```diff
@@ -0,0 +1,254 @@
+        """
+        if not self.outcome and connection.outcome == "BadCertificateEnvironment":
+            self.closing = True
+            self.outcome = connection.outcome
+
+    async def maintain(self):
+        while True:
+            if self.closing or self.closed:
+                return
+
+            for connection in list(self.active):
+                if connection.closing:
+                    self.active.remove(connection)
+                    self.dying.add(connection)
+                    self.termination_hook(connection)
+
+            while len(self.active) > self.size:
+                connection = self.active.pop()
+                connection.closing = True
+                self.dying.add(connection)
+                self.termination_hook(connection)
+
+            for connection in list(self.dying):
+                if connection.closed:
+                    self.dying.remove(connection)
+                    self.termination_hook(connection)
+                elif not connection.channels:
+                    self.dying.remove(connection)
+                    try:
+                        await connection.close()
+                    finally:
+                        self.termination_hook(connection)
+                if self.closing or self.closed:
+                    return
+
+            while len(self.active) < self.size:
+                if not await self.add_one_connection():
+                    break
+                if self.closing or self.closed:
+                    return
+
+            # FIXME wait for a trigger:
+            # * some connection state has changed
+            with suppress(TimeoutError):
+                await wait_for(self.maintenance_needed.wait(), timeout=1)
+            self.maintenance_needed.clear()
+
+    async def add_one_connection(self):
+        try:
+            connection = await Connection.create(self.origin, ssl=self.ssl_context)
+            self.active.add(connection)
+            self.termination_hook(connection)
+            return True
+        except OSError as e:
+            logger.error("%s", e)
+        except Exception:
+            logger.exception("Failed creating APN connection")
+
+    @contextmanager
+    def count_requests(self):
+        try:
+            yield
+        except:
+            self.errors += 1
+            raise
+        else:
+            self.completed += 1
+
+    async def post_once(self, request: "Request") -> "Response":
+        # FIXME ideally, follow weighted round-robin discipline:
+        # * generally allocate requests evenly across connections
+        # * but keep load for few last connections lighter
+        #   to prevent all connections expiring at once
+        # * ideally track connection backlog
+        active = list(self.active)
+        shuffle(active)
+        for connection in active:
+            if self.closing:
+                raise Closed(self.outcome)
+            if connection.closed:
```

Comment:
```suggestion
            if connection.closed or connection.closing:
```

If the connection is already closing, it shouldn't be used right?

Reply:
Done, thanks!